pci/src/msi.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+
+extern crate byteorder;
+extern crate vm_memory;
+
+use byteorder::{ByteOrder, LittleEndian};
+
+// MSI control masks
+const MSI_CTL_ENABLE: u16 = 0x1;
+const MSI_CTL_MULTI_MSG_ENABLE: u16 = 0x70;
+const MSI_CTL_64_BITS: u16 = 0x80;
+const MSI_CTL_PER_VECTOR: u16 = 0x100;
+
+// MSI message offsets
+const MSI_MSG_CTL_OFFSET: u64 = 0x2;
+const MSI_MSG_ADDR_LO_OFFSET: u64 = 0x4;
+
+// MSI message masks
+const MSI_MSG_ADDR_LO_MASK: u32 = 0xffff_fffc;
+
+#[derive(Clone, Copy, Default)]
+pub struct MsiCap {
+    // Message Control Register
+    //   0:     MSI enable.
+    //   3-1;   Multiple message capable.
+    //   6-4:   Multiple message enable.
+    //   7:     64 bits address capable.
+    //   8:     Per-vector masking capable.
+    //   15-9:  Reserved.
+    pub msg_ctl: u16,
+    // Message Address (LSB)
+    //   1-0:  Reserved.
+    //   31-2: Message address.
+    pub msg_addr_lo: u32,
+    // Message Upper Address (MSB)
+    //   31-0: Message address.
+    pub msg_addr_hi: u32,
+    // Message Data
+    //   15-0: Message data.
+    pub msg_data: u16,
+    // Mask Bits
+    //   31-0: Mask bits.
+    pub mask_bits: u32,
+    // Pending Bits
+    //   31-0: Pending bits.
+    pub pending_bits: u32,
+}
+
+impl MsiCap {
+    fn addr_64_bits(&self) -> bool {
+        self.msg_ctl & MSI_CTL_64_BITS == MSI_CTL_64_BITS
+    }
+
+    fn per_vector_mask(&self) -> bool {
+        self.msg_ctl & MSI_CTL_PER_VECTOR == MSI_CTL_PER_VECTOR
+    }
+
+    pub fn enabled(&self) -> bool {
+        self.msg_ctl & MSI_CTL_ENABLE == MSI_CTL_ENABLE
+    }
+
+    pub fn num_enabled_vectors(&self) -> usize {
+        let field = (self.msg_ctl >> 4) & 0x7;
+
+        match field {
+            0b000 => 1,
```

Comment:
Simplify this to `1 << field` ?

Reply:
Oh yes let me address this, I completely miss this simplification :)

Reply:
Done!